### PR TITLE
Don't attempt JAX-RS blocking writes from an IO thread

### LINKED ISSUE #3578: async REST return types will throw when used reactively
Due to the new HTTP-vertx merge, we're starting to see the following exception:

```
2019-08-19 17:24:18,882 ERROR [org.jbo.res.res.i18n] (vert.x-eventloop-thread-30) RESTEASY002020: Unhandled asynchronous exception, sending back 500: java.lang.IllegalStateException: UT000126: Attempted to do blocking IO from the IO thread. This is prohibited as it may result in deadlocks
	at io.undertow.servlet.spec.ServletOutputStreamImpl.write(ServletOutputStreamImpl.java:132)
	at io.undertow.servlet.spec.ServletOutputStreamImpl.write(ServletOutputStreamImpl.java:115)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletResponseWrapper$DeferredOutputStream.write(HttpServletResponseWrapper.java:39)
	at org.jboss.resteasy.util.CommitHeaderOutputStream.write(CommitHeaderOutputStream.java:64)
	at org.jboss.resteasy.plugins.providers.StringTextStar.writeTo(StringTextStar.java:65)
	at org.jboss.resteasy.plugins.providers.StringTextStar.writeTo(StringTextStar.java:24)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.writeTo(AbstractWriterInterceptorContext.java:193)
	at org.jboss.resteasy.core.interception.jaxrs.ServerWriterInterceptorContext.writeTo(ServerWriterInterceptorContext.java:64)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.proceed(AbstractWriterInterceptorContext.java:155)
	at org.jboss.resteasy.core.ServerResponseWriter.lambda$writeNomapResponse$2(ServerResponseWriter.java:156)
	at org.jboss.resteasy.core.interception.jaxrs.ContainerResponseContextImpl.filter(ContainerResponseContextImpl.java:405)
	at org.jboss.resteasy.core.ServerResponseWriter.executeFilters(ServerResponseWriter.java:232)
	at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:97)
	at org.jboss.resteasy.core.AsyncResponseConsumer.sendBuiltResponse(AsyncResponseConsumer.java:148)
	at org.jboss.resteasy.core.AsyncResponseConsumer.internalResume(AsyncResponseConsumer.java:115)
	at org.jboss.resteasy.core.AsyncResponseConsumer$CompletionStageResponseConsumer.accept(AsyncResponseConsumer.java:237)
	at org.jboss.resteasy.core.AsyncResponseConsumer$CompletionStageResponseConsumer.accept(AsyncResponseConsumer.java:216)
	at io.smallrye.context.SmallRyeThreadContext$ContextualBiConsumer.accept(SmallRyeThreadContext.java:157)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at io.smallrye.context.CompletableFutureWrapper.lambda$new$0(CompletableFutureWrapper.java:35)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at io.vertx.axle.AsyncResultCompletionStage.lambda$toCompletionStage$0(AsyncResultCompletionStage.java:18)
	at io.vertx.axle.sqlclient.SqlConnection$3.handle(SqlConnection.java:196)
	at io.vertx.axle.sqlclient.SqlConnection$3.handle(SqlConnection.java:193)
	at io.vertx.sqlclient.impl.SqlResultBuilder.handle(SqlResultBuilder.java:65)
	at io.vertx.sqlclient.impl.SqlResultBuilder.handle(SqlResultBuilder.java:29)
	at io.vertx.sqlclient.impl.SqlConnectionImpl.lambda$schedule$0(SqlConnectionImpl.java:52)
	at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:210)
	at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:81)
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$2(NetSocketImpl.java:101)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:235)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:125)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:364)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:232)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
	at io.vertx.pgclient.impl.codec.PgEncoder.lambda$write$0(PgEncoder.java:77)
	at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:124)
	at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:229)
	at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:87)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```

Which happens when you return something like:

```java
@GET
public CompletionStage<String> foo(){
 return doAsyncHttpCall().thenApply(() -> "OK");
}
```

So, what happens is that we run a client HTTP call, which notifies us of its result on the IO thread, then RESTEasy registers for this async result and proceeds to write the response to the client. 

RESTEasy ATM doesn't care which thread it's invoked on, and will happily resume the suspended HTTP request on the IO thread like any other thread. But Undertow detects this and throws an exception.

This needs fixing, because it will happen a lot. If RESTEasy cannot ever write on the IO thread, it needs to be able to determine that it is on the wrong thread when it gets the async result, and resume the request on another thread (which one??).

OTOH, if @patriot1burke rewrites the `quarkus-resteasy` integration so that it runs without servlet, directly on vert.x, it's possible it will use the `resteasy-vertx` integration which writes all IO as buffered HTTP chunk and so writing the response will likely not care that it's on the IO thread. Though flushing/sending it might still care.

This probably needs advice from @vietj @stuartwdouglas and @cescoffier.

### PR DESCRIPTION
**Pull request by stuartwdouglas**
Instead buffer then write with async IO

Fixes #3578

### CODE REVIEW

**Comment by stuartwdouglas** [comment]
@FroMage can you test this out with your techempower test?

@gsmet It would be good to get this in the release today if it passes and fixes the techempower problem.

**Comment by FroMage** [comment]
Lemme test it right away

File: extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/ResteasyAsyncResponseWrapper.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    private final HttpServletRequest request;
+
+    public ResteasyAsyncResponseWrapper(HttpServletResponse response, HttpServletRequest request) {
+        super(response);
+        this.request = request;
+    }
+
+    @Override
+    public ServletOutputStream getOutputStream() throws IOException {
+        return new ResteasyAsyncOutputStream((HttpServletResponse) getResponse(), request);
```
Comment (FroMage) [inline extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/ResteasyAsyncResponseWrapper.java]: Should this be a new instance at every call?

**Comment by FroMage** [comment]
Confirmed it fixed the issue

**Comment by Sergey80** [comment]
I have this code in some service that is called from the RestEeasy 

  final Flowable<Map<String, List<Data>>> relatedMaps = Flowable.range(0, requestList.size())
                                                                                    .concatMapEager(index ->
                                                                                                    fetchByHttp(requestList.get(index))
                                                                                                            .subscribeOn(Schedulers.io())
                                                                                                            .toFlowable(),
                                                                                            requestList.size() * 2,
                                                                                            1
                                                                                    );

and sometimes I get this issue. Quarkis 1.0 Final.

**Pull request merged**